sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/yaml/YAMLToInternalMappers.java

```diff
@@ -359,12 +361,70 @@ private static DefaultResourceSet from(
         }
 
-        if (rawEndpoints != null) {
-            resourceSetBuilder.addPorts(rawEndpoints);
+        if (rawPorts != null) {
+            resourceSetBuilder.addResource(from(role, principal, rawPorts));
         }
 
         return resourceSetBuilder
                 .id(id)
                 .build();
     }
+
+    private static ResourceSpec from(String role, String principal, WriteOnceLinkedHashMap<String, RawPort> rawPorts) {
+        Collection<PortSpec> portSpecs = new ArrayList<>();
+        Protos.Value.Builder portsValueBuilder = Protos.Value.newBuilder().setType(Protos.Value.Type.RANGES);
+        String envKey = null;
+        for (Map.Entry<String, RawPort> portEntry : rawPorts.entrySet()) {
+            String name = portEntry.getKey();
+            RawPort rawPort = portEntry.getValue();
+            Protos.Value.Builder portValueBuilder = Protos.Value.newBuilder()
+                    .setType(Protos.Value.Type.RANGES);
+            portValueBuilder.getRangesBuilder().addRangeBuilder()
+                    .setBegin(rawPort.getPort())
+                    .setEnd(rawPort.getPort());
+            portsValueBuilder.mergeRanges(portValueBuilder.getRanges());
+            if (envKey == null) {
+                envKey = rawPort.getEnvKey();
+            }
+
+            if (rawPort.getVip() != null) {
+                final RawVip rawVip = rawPort.getVip();
+                final String protocol =
+                        StringUtils.isEmpty(rawVip.getProtocol()) ? DEFAULT_VIP_PROTOCOL : rawVip.getProtocol();
+                final String vipName = StringUtils.isEmpty(rawVip.getPrefix()) ? name : rawVip.getPrefix();
+                portSpecs.add(new NamedVIPSpec(
+                        Constants.PORTS_RESOURCE_TYPE,
+                        portValueBuilder.build(),
+                        role,
+                        principal,
+                        rawPort.getEnvKey(),
+                        name,
+                        protocol,
+                        toVisibility(rawVip.isAdvertised()),
+                        vipName,
+                        rawVip.getPort()));
+            } else {
+                portSpecs.add(new PortSpec(
+                        Constants.PORTS_RESOURCE_TYPE,
+                        portValueBuilder.build(),
+                        role,
+                        principal,
+                        rawPort.getEnvKey(),
+                        name));
+            }
+        }
+        return new PortsSpec(
+                Constants.PORTS_RESOURCE_TYPE, portsValueBuilder.build(), role, principal, envKey, portSpecs);
+    }
+
+    /**
+     * This visibility information isn't currently used by DC/OS Service Discovery. At the moment it's only enforced in
+     * our own {@link com.mesosphere.sdk.api.EndpointsResource}.
+     */
+    private static DiscoveryInfo.Visibility toVisibility(Boolean rawIsVisible) {
+        if (rawIsVisible == null) {
```

Comment:
Other `Raw*` classes, such as `RawPod` do null conversions in the constructor, so we can get rid of null values early.  Maybe we should do that here.  Then `PUBLIC_VIP_VISIBILITY` can go in `RawVIP`.  And it should probably be called `DEFAULT_VIP_VISIBILITY`.  That's my understanding of what it is.